src/utils/buildActivityCsvs.ts

```diff
@@ -0,0 +1,517 @@
+  }
+}
+
+export async function downloadRedemptions(
+  blockNumber: number | undefined,
+  projectId: number | undefined,
+  pv: PV | undefined,
+) {
+  if (blockNumber === undefined || !projectId || !pv) return
+
+  const rows = [
+    [
+      t`Date`,
+      t`Tokens redeemed`,
+      t`ETH received`,
+      t`USD value of ETH received`,
+      t`Caller`,
+      t`Beneficiary`,
+      t`Transaction hash`,
+    ], // CSV header row
+  ]
+
+  try {
+    const redemptions = await querySubgraphExhaustive({
+      entity: 'redeemEvent',
+      keys: [
+        'timestamp',
+        'amount',
+        'returnAmount',
+        'returnAmountUSD',
+        'caller',
+        'beneficiary',
+        'txHash',
+      ],
+      orderBy: 'timestamp',
+      orderDirection: 'desc',
+      block: {
+        number: blockNumber,
+      },
+      where: [
+        {
+          key: 'projectId',
+          value: projectId,
+        },
+        {
+          key: 'pv',
+          value: pv,
+        },
+      ],
+    })
+
+    if (!redemptions) {
+      emitErrorNotification(t`Error loading redemptions`)
+      throw new Error('No data.')
+    }
+
+    redemptions.forEach(r => {
+      let date = new Date((r.timestamp ?? 0) * 1000).toUTCString()
+
+      if (date.includes(', ')) date = date.split(', ')[1]
+
+      rows.push([
+        date,
+        fromWad(r.amount),
+        fromWad(r.returnAmount),
+        fromWad(r.returnAmountUSD),
+        r.caller,
+        r.beneficiary,
+        r.txHash,
+      ])
+    })
+
+    downloadCsvFile(
+      `redemptions_v${pv}p${projectId}_block-${blockNumber}`,
+      rows,
+    )
+  } catch (e) {
+    console.error('Error downloading redemptions', e)
+    emitErrorNotification(t`Error downloading redemptions, try again.`)
+  }
+}
+
+export async function downloadAdditionsToBalance(
+  blockNumber: number | undefined,
+  projectId: number | undefined,
+  pv: PV | undefined,
+) {
+  if (blockNumber === undefined || !projectId || !pv) return
+
+  const rows = [
+    [
+      t`Date`,
+      t`ETH added`,
+      t`USD value of ETH added`,
+      t`Caller`,
+      t`Transaction hash`,
+    ], // CSV header row
+  ]
+
+  try {
+    const additions = await querySubgraphExhaustive({
+      entity: 'addToBalanceEvent',
+      keys: ['timestamp', 'amount', 'amountUSD', 'caller', 'txHash'],
+      orderBy: 'timestamp',
+      orderDirection: 'desc',
+      block: {
+        number: blockNumber,
+      },
+      where: [
+        {
+          key: 'projectId',
+          value: projectId,
+        },
+        {
+          key: 'pv',
+          value: pv,
+        },
+      ],
+    })
+
+    if (!additions) {
+      emitErrorNotification(t`Error loading additions to balance`)
+      throw new Error('No data.')
+    }
+
+    additions.forEach(a => {
+      let date = new Date((a.timestamp ?? 0) * 1000).toUTCString()
+
+      if (date.includes(', ')) date = date.split(', ')[1]
```

Comment:
> **suggestion** looks like this code can be consolidated into a single util to reduce duplication. WDYT?

Up to you! There are a few things which could be broken up into utils but it would make the code a bit harder to follow. Worth keeping consolidated in the functions IMO.